Allow default args of 0

gchenfc opened the pull request:
Currently the check in `pybind_wrapper.py` for whether there is a default arg or not is just
```python
  if ...:
    # deal with string manually
   ...
  elif arg.default:
    # yes default arg
    ...
  else:
    # no default arg
    ...
```

But since things like `0` and `""` are False-y, then this check misses those cases.  Should be `elif arg.default is not None:`.

ProfFan commented:
Need a unit test :)

varunagrawal commented:
Nice catch, but @ProfFan is right. Needs a dedicated unit test for both interface_parser and pybind_wrapper.

gchenfc commented:
K, I'll add one for `interface_parser`.  Is the one I had added for `pybind_wrapper` good enough or do you want me to add more for that too?

gchenfc commented:
I realized that `identifier`s are also not being parsed correctly (they're getting wrapped into strings) so I added a unit test for that.

Then I changed the behavior of quoted string parsing so that it doesn't automatically unquote them.  I believe this is advantageous for two reasons:
1. We don't have to manually deal with string literals
2. otherwise `char c = 'a'` would get changed to `char c = "a"` which I believe is not correct behavior.  Added a unit test for this as well.

Please let me know if you disagree and I can revert those and just have the minimum code needed for dealing with '0' in this PR.  Otherwise, I think this is a well-justified change.

varunagrawal commented on tests/test_interface_parser.py:
```diff
@@ -181,29 +181,34 @@ def test_argument_list_templated(self):
-            "gtsam::KeyFormatter kf = gtsam::DefaultKeyFormatter, "
-            "std::vector<size_t> p = std::vector<size_t>(), "
-            "std::vector<size_t> l = (1, 2, 'name', \"random\", 3.1415)"
+            """string c = "", int z = 0, double z2 = 0.0, bool f = false, """
```
Simply use backslashes rather than add all those extra quotes? There's no rule that says we can't use escape characters when needed.

gchenfc replied:
You're right.  Replaced with a single multiline string, how's that?

I just find escaped characters difficult to read, which probably I shouldn't have changed code based on personal preference, but now with multiline comment I think that's actually maybe the most appropriate and readable?

varunagrawal commented on tests/test_interface_parser.py:
```diff
@@ -181,29 +181,34 @@ def test_argument_list_templated(self):
+            """int b, double pi = 3.1415, """
+            """gtsam::KeyFormatter kf = gtsam::DefaultKeyFormatter, """
+            """std::vector<size_t> p = std::vector<size_t>(), """
+            """std::vector<size_t> l = (1, 2, "name", "random", 3.1415)"""
```
Please keep the single and double quote tests. I put those on purpose to test for both cases.

gchenfc replied:
done. I had assumed `char c = 'a'` replaced `"name"` but I've also added a char literal inside the tuple.

varunagrawal reviewed (approved):
Small requests.

varunagrawal reviewed (changes requested):
Please re-request a review once you've addressed the comments?

varunagrawal commented:
> I realized that `identifier`s are also not being parsed correctly (they're getting wrapped into strings) so I added a unit test for that.

I believe this is pyparsing's default behavior?

> 1. We don't have to manually deal with string literals

An example of this? Is it like the case of `bool flag = false`? Asking so I have the complete picture.

> 2. otherwise `char c = 'a'` would get changed to `char c = "a"` which I believe is not correct behavior.  Added a unit test for this as well.

Yup, you're right.

gchenfc commented:
> > I realized that `identifier`s are also not being parsed correctly (they're getting wrapped into strings) so I added a unit test for that.
> 
> I believe this is pyparsing's default behavior?

Yes, I should have said that _we_ are not parsing identifiers properly since string literals have their quotes removed so that they are indistinguishable from identifiers (non-quoted literals), e.g. false, below

> 
> > 1. We don't have to manually deal with string literals
> 
> An example of this? Is it like the case of `bool flag = false`? Asking so I have the complete picture.
> 

Yes, currently `bool flag = false` gets parsed as a string `'false'`, but the problem is that currently `pybind_wrapper` treats all strings as string literals and will automatically add double quotes around `false`, so it parses this as a string literal rather than an identifier.  Then in the generated cpp file, it will generate `bool flag = "false"` which is not correct.  This could conceivably occur for variable names as well, but I'm not sure when a variable would be not inside namespace (inside a namespace it will be caught by "Type" rather than by "DEFAULT_ARG").

varunagrawal reviewed (approved):
LGTM
